File: internal/translator/formats.go
// Package translator provides core translator registries, types, and format detection logic.
package translator

import "strings"

// Supported translator format identifiers.
const (
	FormatOpenAI          = "openai"
	FormatOpenAIResponses = "openai-responses"
	FormatOpenAIResponse  = "openai-response"
	FormatClaude          = "claude"
	FormatGemini          = "gemini"
	FormatGeminiCLI       = "gemini-cli"
	FormatVertex          = "vertex"
	FormatCodex           = "codex"
	FormatAntigravity     = "antigravity"
	FormatKiro            = "kiro"
	FormatCursor          = "cursor"
	FormatOllama          = "ollama"
	FormatCommandCode     = "commandcode"
	FormatResponses       = "responses"
)

// DetectFormatByEndpoint detects the API format based on request pathname and body.
func DetectFormatByEndpoint(pathname string, body map[string]any) string {
	if strings.Contains(pathname, "/v1/responses") {
		return FormatOpenAIResponses
	}

	if strings.Contains(pathname, "/v1/messages") {
		return FormatClaude
	}

	if strings.Contains(pathname, "/v1/chat/completions") {
		if _, ok := body["input"]; ok {
			if _, isArray := body["input"].([]any); isArray {
				return FormatOpenAI
			}
		}
	}

	return ""
}

// DetectSourceFormat inspects the request body payload to determine its source format.
func DetectSourceFormat(body map[string]any) string {
	if _, ok := body["thinking"]; ok {
		return FormatClaude
	}

	if _, ok := body["system"]; ok {
		return FormatClaude
	}

	if _, ok := body["contents"]; ok {
		return FormatGemini
	}

	if _, ok := body["generationConfig"]; ok {
		return FormatGemini
	}

	if _, ok := body["input"]; ok {
		return FormatOpenAIResponses
	}

	return FormatOpenAI
}

File: open-sse/src/formats.rs
//! Format identifiers — mirror of open-sse/translator/formats.js

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Format {
    #[serde(rename = "openai")]
    Openai,
    #[serde(rename = "openai-responses")]
    OpenaiResponses,
    #[serde(rename = "openai-response")]
    OpenaiResponse,
    #[serde(rename = "claude")]
    Claude,
    #[serde(rename = "gemini")]
    Gemini,
    #[serde(rename = "gemini-cli")]
    GeminiCli,
    #[serde(rename = "vertex")]
    Vertex,
    #[serde(rename = "codex")]
    Codex,
    #[serde(rename = "antigravity")]
    Antigravity,
    #[serde(rename = "kiro")]
    Kiro,
    #[serde(rename = "cursor")]
    Cursor,
    #[serde(rename = "ollama")]
    Ollama,
    #[serde(rename = "commandcode")]
    Commandcode,
}

impl Format {
    pub fn as_str(&self) -> &'static str {
        match self {
            Format::Openai => "openai",
            Format::OpenaiResponses => "openai-responses",
            Format::OpenaiResponse => "openai-response",
            Format::Claude => "claude",
            Format::Gemini => "gemini",
            Format::GeminiCli => "gemini-cli",
            Format::Vertex => "vertex",
            Format::Codex => "codex",
            Format::Antigravity => "antigravity",
            Format::Kiro => "kiro",
            Format::Cursor => "cursor",
            Format::Ollama => "ollama",
            Format::Commandcode => "commandcode",
        }
    }
}

/// Mirror of detectFormatByEndpoint in open-sse/translator/formats.js
pub fn detect_format_by_endpoint(pathname: &str, body: &serde_json::Value) -> Option<Format> {
    if pathname.contains("/v1/responses") {
        return Some(Format::OpenaiResponses);
    }
    if pathname.contains("/v1/messages") {
        return Some(Format::Claude);
    }
    if pathname.contains("/v1/chat/completions")
        && body.get("input").map(|v| v.is_array()).unwrap_or(false)
    {
        return Some(Format::Openai);
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn endpoint_detect() {
        assert_eq!(
            detect_format_by_endpoint("/v1/messages", &json!({})),
            Some(Format::Claude)
        );
        assert_eq!(
            detect_format_by_endpoint("/v1/responses", &json!({})),
            Some(Format::OpenaiResponses)
        );
        assert_eq!(
            detect_format_by_endpoint("/v1/chat/completions", &json!({"input": []})),
            Some(Format::Openai)
        );
        assert_eq!(
            detect_format_by_endpoint("/v1/chat/completions", &json!({})),
            None
        );
    }
}
